docs/sequence.rst

```diff
@@ -0,0 +1,647 @@
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Create an empty sequence
+""""""""""""""""""""""""
+
+::
+
+    Seq.empty :: Seq a
+    Seq.empty = ...
+
+:seq:`empty` creates a sequence with zero elements.
+
+::
+
+    Seq.empty
+    > fromList []
+
+
+Create a sequence with one element (singleton)
+""""""""""""""""""""""""""""""""""""""""""""""
+
+::
+
+    Seq.singleton :: a -> Seq a
+    Seq.singleton x = ...
+
+:seq:`singleton` creates a sequence with the single element ``x`` in it.
+
+::
+
+    Seq.singleton "containers"
+    > fromList ["containers"]
+
+    Seq.singleton 1
+    > fromList [1]
+
+Create a sequence with the same element repeated
+""""""""""""""""""""""""""""""""""""""""""""""""
+
+::
+
+    Seq.replicate :: Int -> a -> Seq a
+    Seq.replicate n x = ...
+
+:seq:`replicate` creates a sequence with same element ``x`` repeated ``n``
+times.
+
+::
+
+    Seq.replicate 0 "hi"
+    > fromList []
+
+    Seq.replicate 3 "hi"
+    > fromList ["hi","hi","hi"]
+
+Create a sequence from a list
+"""""""""""""""""""""""""""""
+
+::
+
+    Seq.fromList :: [a] -> Seq a
+    Seq.FromList xs = ...
+
+:seq:`fromList` creates a sequence containing the elements of the list
+``xs``. Sequences allow duplicate so all elements will be included in the order
+given.
+
+::
+
+    Seq.fromList ["base", "containers", "QuickCheck"]
+    > fromList ["base","containers","QuickCheck"]
+
+    Seq.fromList [0, 1, 1, 2, 3, 1]
+    > fromList [0,1,1,2,3,1]
+
+Adding to an existing sequence
+""""""""""""""""""""""""""""""
+
+::
+
+    (<|) :: a -> Seq a -> Seq a
+    x <| xs = ...
+
+    (|>) :: Seq a -> a -> Seq a
+    xs |> x = ...
+
+    (><) :: Seq a -> Seq a -> Seq a
+    l >< r = ...
+
+- ``x <| xs`` places the element ``x`` at the beginning of the sequence ``xs``..
+
+- ``xs |> x`` places the element ``x`` at the end of the sequence ``xs``.
+
+- ``l >< r`` combines the two sequences ``l`` and ``r`` together.
+
+
+Create a list from a sequence
+"""""""""""""""""""""""""""""
+
+::
+
+    foldr (:) [] :: Seq a -> [a]
+
+
+There is no ``toList`` function in the Sequence module since it can be
+`easily implemented <https://wiki.haskell.org/Foldable_and_Traversable>`_ with a
+fold using ``Seq``'s `Foldable
+<https://wiki.haskell.org/Typeclassopedia#Foldable>`_ instance.
+
+::
+
+    foldr (:) [] (Seq.fromList ["base", "containers", "QuickCheck"])
+    > ["base","containers","QuickCheck"]
+
+
+Pattern Matching
+^^^^^^^^^^^^^^^^
+
+*Since 0.5.10*
+
+Just like you can pattern match (aka. destructure) a list ``[a]``, you can do
+the same with sequneces. Lets first look at how we do this with lists::
+
+    case [1, 2, 3] of
+      [] -> "empty list"
+      (x:xs) -> "first:" ++ show x ++ " rest:" ++ show xs
+    > "first:1 rest:[2,3]"
+
+
+We can do the same thing with sequences, but first we'll need to enable the
+``PatternSynonyms`` language extension. At the top of your Haskell source file
+above the ``module`` declaration add::
+
+   {-# LANGUAGE PatternSynonyms #-}
```

Comment:
The language extension should not be required to *use* the patterns with recent GHC versions. Just use

```haskell
import Data.Sequence (Seq (..))
```
or
```haskell
import Data.Sequence (Seq (Empty, (:<|), (:|>)))
```

rather than the older (and `PatternSynonyms`-only) syntax

```haskell
import Data.Sequence (pattern (:<), pattern (:>), pattern Empty)
```

Reply:
TIL, thanks!